File: Examples/TypeScript/Containers/NodeJS-MongoDB-REST/src/controllers/health.controller.ts
import { Request, Response } from 'express';
import MongoDB_Config from '../services/MongoDB/CRUD/mongodb-config';
import MongoDB_CRUD_Client from '../services/MongoDB/CRUD/mongodb-crud-client';
import MongoDB_REST_Client from '../services/MongoDB/REST/mongodb-rest-client';
import { HealthResultEntry } from '../health/health-result-entry.model';
import { HealthResults } from '../health/health-results.model';

export class Health_Controller {
    constructor() {
        this.get = this.get.bind(this);
    }

    private createMongoDBClient(): MongoDB_REST_Client {
        return new MongoDB_REST_Client(new MongoDB_CRUD_Client());
    }

    private async builHealthCheckMongoDB(mongodb_rest_client: MongoDB_REST_Client): Promise<HealthResultEntry> {
        const start = Date.now();
        try {
            await mongodb_rest_client.ping();
            return new HealthResultEntry('mongodb', true, Date.now() - start);
        } catch (e) {
            return new HealthResultEntry('mongodb', false, Date.now() - start, (e as Error).message);
        }
    }
    
    private async buildHealthCheckResponse(mongodb_rest_client: MongoDB_REST_Client): Promise<HealthResults> {
        const start = Date.now();
        const results = [await this.builHealthCheckMongoDB(mongodb_rest_client)]
        const duration = Date.now() - start;
        return new HealthResults(duration, results)
    }

    async get(req: Request, res: Response) {
        const client = this.createMongoDBClient();
        const response = await this.buildHealthCheckResponse(client);
        res.status(response.healthy ? 200 : 500).json(response);
    }
}

File: Examples/Java/Containers/MongoDB-REST/src/main/java/com/api/mongodb/rest/controller/HealthController.java
package com.api.mongodb.rest.controller;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.stream.Collectors;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import com.api.mongodb.rest.health.HealthResultEntry;
import com.api.mongodb.rest.health.HealthResults;
import com.api.mongodb.rest.service.mongodb.crud.*;
import com.api.mongodb.rest.service.mongodb.rest.*;

@RestController
@RequestMapping("/api/health")
public class HealthController {
    private MongoDB_REST_Client createMongoDBClient() {
        return new MongoDB_REST_Client(new MongoDB_CRUD_Client());
    }

    private CompletableFuture<HealthResultEntry> buildHealthCheckMongoDB(MongoDB_REST_Client mongodb_rest_client) {
        long start = System.currentTimeMillis();
        return mongodb_rest_client.ping()
            .thenApply(v -> new HealthResultEntry("mongodb", true, System.currentTimeMillis() - start, null))
            .exceptionally(e-> new HealthResultEntry("mongob", false, System.currentTimeMillis() - start, e.getMessage()));
    }

    private CompletableFuture<HealthResults> buildHealthCheckResponse(MongoDB_REST_Client mongodb_rest_client) {
        long start = System.currentTimeMillis();
        List<CompletableFuture<HealthResultEntry>> futures = new ArrayList<>();
        futures.add(buildHealthCheckMongoDB(mongodb_rest_client));
        CompletableFuture<Void> allOf = CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]));
        return allOf
            .thenApply(v ->  {
                List<HealthResultEntry> results = futures.stream()
                    .map(CompletableFuture::join)
                    .collect(Collectors.toList());
                return new HealthResults(System.currentTimeMillis() - start, results);
            });
    }

    @GetMapping()
    @PostMapping()
    public ResponseEntity<HealthResults> getHealth() throws InterruptedException, ExecutionException {
        MongoDB_REST_Client client = createMongoDBClient();
        HealthResults healthResults = buildHealthCheckResponse(client).get();
        return ResponseEntity.status(healthResults.isHealthy() ? 200 : 500).body(healthResults);
    }
}